### DISCUSSION TITLE
Fixing Leipzig corpus

### DISCUSSION THREAD

**Discussion by iandoug** [binarybottle/engram-es#17] [Ideas]
Declutter threads.

From "known words" list:
internacional	157568
internacionalice	14
internacionalicemos	4
internacionalicen	6
internacionalidad	107
internacionalismo	291
internacionalista	610
internacionaliza	26
internacionalizaba	2
internacionalización	1898
internacionalizado	63
internacionalizamos	1
internacionalizan	11
internacionalizando	22
internacionalizar	232
internacionalizara	1
internacionalizaran	1
internacionalizaremos	1
internacionalizaron	11
internacionalizará	2
internacionalizaría	1
internacionalizó	42
internacionalmente	3514

From "unknown words" list: Are these (generally) supposed to be one word? Am trying to differentiate between  "normal usage" and badly collected/cleaned  text.

internacionala	4
internacionalactualización	2
internacionalalex	1
internacionalallan	1
internacionalamente	1
internacionalamerica	1
internacionalamericanaaustraliana	1
internacionalamérica	1
internacionalanimadora	1
internacionalargentina	1
internacionalargentino	1
internacionalarévalo	1
internacionalasdi	2
internacionalasistencia	1
internacionalasí	1
internacionalatentados	1
internacionalatlético	1
internacionalaunque	2
internacionalavisos	1
internacionalavtocomplex	1
internacionalbajo	1
internacionalbarcelona	1
internacionalbueno	1
internacionalcampeonato	1
internacionalcapítulo	1
internacionalcbi	1
internacionalch	1
internacionalchile	2
internacionalchilenos	4
internacionalcolombia	1
internacionalcomo	1
internacionalcon	3
internacionalconsecuencias	1
internacionalconsultor	1
internacionalcontactos	1
internacionalcoritiba	1
internacionalcruz	1
internacionalcti	1
internacionalcuando	1
internacionalcumbre	2
internacionalcuál	1
internacionalcáffaro	1
internacionald	1
internacionaldaniele	1
internacionalde	6
internacionaldeberia	1
internacionaldebido	2
internacionaldeclaran	1
internacionaldefinen	1
internacionaldenunciará	1
internacionaldesde	1
internacionaldetalla	1
internacionaldiario	1
internacionaldivisión	1
internacionaldocentes	1
internacionaldt	1
internacionaldónde	1
internacionale	10
internacionaleden	1
internacionaleei	1
internacionalees	1
internacionalel	17
internacionaleliminatorias	5
internacionalels	1
internacionalemente	4
internacionalemilio	1
internacionalemnte	1
internacionalemtne	1
internacionalen	6
internacionalenfrentar	1
internacionalentrenador	1
internacionalequipo	1
internacionalerp	1
internacionales	57489
internacionalesa	1
internacionalesactualmente	1
internacionalesaguas	1
internacionalesapuntan	1
internacionalesarzobispo	1
internacionalesasi	1
internacionalesatentados	1
internacionalesbank	1
internacionalesbogotá	1
internacionalesbolivia	1
internacionalesbush	1
internacionalescanciller	1
internacionalescarlos	1
internacionalescarrá	1
internacionalescasi	1
internacionalescelebran	1
internacionaleschina	1
internacionaleschávez	1
internacionalesclasificados	1
internacionalescolombianos	1
internacionalescomienza	1
internacionalescondecora	1
internacionalesconfirman	1
internacionalesconflicto	1
internacionalescontinúa	1
internacionalescrucial	1
internacionalescue	1
internacionalesde	2
internacionalesdebates	1
internacionalesdiosdado	1
internacionalesdirección	1
internacionalesdos	1
internacionalesdía	1
internacionaleseconómicos	1
internacionalesecuador	1
internacionaleseditar	1
internacionalesedward	1
internacionaleseeuu	1
internacionalesejército	1
internacionalesel	5
internacionaleseligen	1
internacionalesemilio	1
internacionalesempresa	1
internacionalesen	1
internacionalesentre	1
internacionaleses	1
internacionalesexigirán	1
internacionalesfavor	1
internacionalesfidel	1
internacionalesfin	1
internacionalesforo	1
internacionalesfrancia	1
internacionalesfuerzas	1
internacionalesgobierno	1
internacionaleshija	1
internacionaleshomenajean	1
internacionaleshttpwww	1
internacionalesintensifican	1
internacionalesinter	1
internacionalesinternas	1
internacionalesintervención	1
internacionalesinvita	1
internacionalesjornada	1
internacionalesla	5
internacionaleslas	2
internacionalesldquopero	1
internacionaleslondres	1
internacionalesmaria	1
internacionalesmartes	2
internacionalesmerkel	1
internacionalesmoscu	1
internacionalesnadie	1
internacionalesnorcorea	1
internacionalesobama	1
internacionalesola	1
internacionalesonu	1
internacionalespanama	1
internacionalespara	1
internacionalesparlamento	1
internacionalespor	1
internacionalesportal	1
internacionalespresidente	3
internacionalespresionan	1
internacionalesprimer	2
internacionalesprocurando	1
internacionalesque	1
internacionalesquinta	1
internacionalesrealizan	1
internacionalesreanudaran	1
internacionalesreelección	1
internacionalesretiran	1
internacionalesrodriguezunju	1
internacionaless	1
internacionalessanto	3
internacionalessin	1
internacionalessolo	1
internacionalestambién	1
internacionalesteban	1
internacionalesto	1
internacionalestrella	1
internacionalestrump	1
internacionalestá	1
internacionalesuniversidad	1
internacionalesunsam	1
internacionalesvenezuela	1
internacionalesverifica	1
internacionalesvie	1
internacionalesy	1
internacionales¨	1
internacionalexige	1
internacionalexigen	1
internacionalfigo	1
internacionalfmi	4
internacionalfue	1
internacionalfundación	1
internacionalgareca	1
internacionalgoles	1
internacionalgran	1
internacionalgremio	1
internacionalguardiola	2
internacionalha	1
internacionalhay	1
internacionalhelicobacter	1
internacionalheynckes	1
internacionalhrw	1
internacionalhttpst	1
internacionalhttpwww	1
internacionali	1
internacionaliación	1
internacionalicardi	1
internacionalidades	89
internacionalifi	1
internacionalimagen	1
internacionalindia	1
internacionalinflación	1
internacionalinicia	1
internacionalinstituto	1
internacionalinteligencia	1
internacionalinternacional	1
internacionalinterpol	1
internacionalinvitado	1
internacionalis	3
internacionalisaac	1
internacionalismohttpwww	1
internacionalistaantiglobalización	1
internacionalistala	1
internacionalistaliga	1
internacionalistasolidaridad	3
internacionalius	1
internacionalivan	2
internacionalizaci	15
internacionalizacin	19
internacionalizacioacuten	1
internacionalizacion	2
internacionalizaciones	1
internacionalizació	1
internacionalizacióncolocandose	1
internacionalizaciónlan	1
internacionalizada	28
internacionalizadas	38
internacionalizador	1
internacionalizadora	1
internacionalizarla	3
internacionalizarlas	1
internacionalizarlo	6
internacionalizarme	3
internacionalizarnos	3
internacionalizarse	64
internacionalizarsey	1
internacionalize	1
internacionalizándose	3
internacionaljefe	1
internacionalk	1
internacionalla	20
internacionallas	3
internacionalle	2
internacionalledecky	1
internacionallejos	1
internacionalleyenda	2
internacionalllegando	1
internacionallo	3
internacionallocal	1
internacionallorenzo	1
internacionallos	1
internacionalluego	1
internacionalmaradona	1
internacionalmarketing	1
internacionalmaría	1
internacionalmayores	1
internacionalmenteacoge	1
internacionalmentede	1
internacionalmentedesde	1
internacionalmenteen	1
internacionalmenteinditex	1
internacionalmenteno	1
internacionalmento	1
internacionalmete	1
internacionalmisión	1
internacionalmourinho	3
internacionalmujer	1
internacionalmundial	1
internacionalmuralista	1
internacionalnacional	1
internacionalneymar	1
internacionalno	1
internacionalnos	1
internacionalobstaculizar	1
internacionaloea	1
internacionalopositores	1
internacionalordóñez	1
internacionalpara	2
internacionalparaliza	1
internacionalpaula	1
internacionalpelé	1
internacionalpermanecidas	1
internacionalpero	1
internacionalpochettino	1
internacionalpopulismo	1
internacionalpor	1
internacionalpreguntas	38
internacionalproblema	1
internacionalpronóstico	1
internacionalpágina	1
internacionalque	4
internacionalqueda	1
internacionalqué	1
internacionalramón	1
internacionalreal	1
internacionalrecibiría	1
internacionalreforzarán	1
internacionalregión	1
internacionalren	1
internacionalreporteros	1
internacionalreunión	1
internacionalribéry	1
internacionalriverboca	1
internacionalrueda	1
internacionalrumbo	13
internacionalsan	1
internacionalsanta	6
internacionalsantiago	1
internacionalse	2
internacionalsean	1
internacionalsección	1
internacionalselecciones	4
internacionalselección	3
internacionalsevilla	1
internacionalsi	1
internacionalsimeone	1
internacionalsin	1
internacionalsinceridad	1
internacionalsino	1
internacionalsorprendido	1
internacionalsúper	1
internacionalteléfono	1
internacionaltoutain	1
internacionaltreinta	1
internacionaltu	1
internacionalun	3
internacionalusted	1
internacionalvarios	2
internacionalvehículos	1
internacionalvenezuela	2
internacionalvideochatinicio	219
internacionalviernes	1
internacionalvulneración	1
internacionalyo	1
internacionalzación	1
internacionalísimo	1

Thanks, Ian

**Comment by Lobo-Feroz** [comment]
The words in the second list seem to be two concatenated words, missing the space in between:

internacionalactualización --> "internacional actualización" (international update)

Some are typos:

internacionalzación --> "internacional**i**zación"

BTW I thought that maybe you could add the grave and circumflex accents to your list of characters: àèìòù ÀÈÌÒÙ âêîôû ÂÊÎÔÛ

While they are not used in spanish, they could appear somewhere like for instance the literal name of a frech person's name, location or movie, like with english.

**Comment by iandoug** [comment]
Can't think of an easy/fast way to figure out if it is a concatenated word. It is doable, but not fast.

I will have to ignore them. Will make some program changes to try to find more plurals, and then see what is fixable and what must be ignored.

My list of characters is going to be the "baseline" that any keyboard layout MUST provide if it claims to be "Spanish". On the flip side, all input texts may only contain those characters, so that texts and keyboards are in agreement. Other characters are optional.

**Reply by Lobo-Feroz** [reply]
> Can't think of an easy/fast way to figure out if it is a concatenated word. It is doable, but not fast.

How many are we talking about? I think I can manually clean a txt file of that rather quickly (it's basically hitting space).

> My list of characters is going to be the "baseline" that any keyboard layout MUST provide if it claims to be "Spanish". On the flip side, all input texts may only contain those characters, so that texts and keyboards are in agreement. Other characters are optional.

Ah, ok, with the goal in mind, it's clearer. 

In that case, I think you should drop these two symbols: § ¶ 

Technically, they are spanish punctuation symbols. But they are not on qwerty-es, and noone, ever, uses them in any kind of text.

The frequencies are negligible as well:

```
§	332	1.0E-5
¶	124	0
```

Also the yen and pound are not in qwerty-es. I think they're useful. but they don't follow your rule of being in a spanish keyboard. But I think in an optimized kb it'd be nice to have them: AlGr + L for £, and AltGr + Y for ¥.

Also, I wonder why `# = [ ] { } _ |` have a frequency of zero. These are valid symbols, and they have their uses...

**Reply by iandoug** [reply]
> In that case, I think you should drop these two symbols: § ¶

Okay, that also reduces bigram list a lot.
> 
> Also the yen and pound are not in qwerty-es. I think they're useful. but they don't follow your rule of being in a spanish keyboard. But I think in an optimized kb it'd be nice to have them: AlGr + L for £, and AltGr + Y for ¥.
> 
> Also, I wonder why `# = [ ] { } _ |` have a frequency of zero. These are valid symbols, and they have their uses...

Those are not in the Leipzig corpus ( I scanned for some of them to convince myself.). could be because of their sources ... If I get Gutenberg to work some should show up. I had similar issues with the English corpus that I made (no Leipzig).

Current "Spanish" character list is 
`
abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890~!@#$%^&*()-_=+\|]}[{"';:/?.>,<«»¨´§·¿¡ºª€¶¬áéíóúüñçÁÉÍÓÚÜÑÇ‘’“”`
`
Will remove §	and ¶.
Typographic quotes are at the end. Github is taking the backtick as "end of code" marker so it's on the last line

**Reply by Lobo-Feroz** [reply]
``abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890~!@#$%^&*()-_=+\|]}[{"';:/?.>,<«»¨´§·¿¡ºª€¶¬áéíóúüñçÁÉÍÓÚÜÑÇ‘’“”` ``

You can use MOAR backticks :)

Here I opened with 2 backticks, closed with two backticks.

The escaped backtick shall not be next to the code backticks, so I added a space at the end.

https://github.com/github/markup/issues/363

**Comment by iandoug** [comment]
Are these actual words?
I have made progress cleaning .... add lots of Capitalised words to the 'known words' list. These are either nouns, acronyms, or the start of sentences, but all were there at least 9 times so assuming okay.

Here's the top 100 of the others... I'm guessing they are all okay. As you can see the counts are getting low.  
Metrics are improving:
Found 5,529,457 bad lines out of 26,329,994 lines
Found 1,642,613 distinct bad words and 612,210 distinct good words  
I don't expect anyone to work through a list of 1.64 million words ... :-)

éste	52797
ésta	35450
órdenes	18416
éstos	15970
éstas	11552
ó	9910
inmediaciones	8537
tratarse	7027
considerarse	5358
comprension	5065
sinonimo	5049
producirse	5018
desarrollarse	4644
imponerse	4324
antonimo	4251
generada	4120
é	4093
cumplirse	3797
decirse	3777
utilizarse	3681
involucradas	3653
usarse	3645
aplicarse	3582
ése	3104
celebrarse	3096
referida	2998
marcharse	2842
ocurrida	2761
ubicarse	2639
óptima	2593
extenderse	2578
establecerse	2570
pronunciarse	2529
priísta	2445
aledañas	2416
drásticamente	2388
involucrada	2189
concretarse	2165
diciéndole	2143
ésa	2124
futbolística	2095
reflejada	2091
entenderse	2083
haberla	2081
perredista	2032
resolverse	2022
jugarse	2020
ayudarlo	2010
lesionadas	1960
ósea	1952
repetirse	1935
matarlo	1922
infectadas	1889
robarle	1867
generadas	1861
explícita	1861
confirmarse	1853
definirse	1837
pagarle	1828
masivamente	1820
óptimas	1815
utilizarlo	1810
construirse	1802
graduarse	1797
contraparte	1797
permitirse	1785
referidas	1773
ícono	1760
dejarla	1756
esconderse	1743
vengarse	1720
errónea	1719
anotada	1702
buscarlo	1669
exmandatario	1664
mostrarle	1655
esperarse	1637
disputarse	1629
evacuadas	1628
salvarse	1622
contaminada	1616
entregarle	1595
étnica	1582
aprobarse	1564
observarse	1563
homónima	1561
ayudarlos	1537
considerarlo	1525
expresada	1520
reducirse	1510
mismísimo	1508
pagada	1494
acogerse	1488
investigada	1486
democráticamente	1484
declararse	1479
ésto	1472
á	1460
ayudarte	1457

Thanks, Ian

**Reply by Lobo-Feroz** [reply]
The majority of those are correct words. Perhaps a little on the rare side, meaning they're not very common, and maybe that's why they aren't in the dictionary. But they are fine.

Some have typos, mostly with the accented vowels. Let me try to catch them all:

comprension --> comprensión (understanding)
ésto --> esto (this)
sinonimo --> sinónimo
antonimo --> antónimo
é --> e (as a standalone word, 'e' means '&', but it's not accented)
priísta --> priista
á --> a (same as 'e', as a standalone word, 'a' means 'to', but it's not accented)

A special case is "ícono". The typical form is "icono" (unaccented) but I just checked and accented is also correct. It's more likely a typo, but it happens to be valid.

I think those would be all, if I haven't missed any.

**Reply by iandoug** [reply]
Thanks. I guess I will use regex to replace them in the corpus.
I need to figure out how to rewrite the corpus ... To examine the words I had to first strip all the fluff (punctuation etc) away, but when I rewrite to throw away all the low-use-maybe-typo and other junk, I will need to keep all the punctuation. Will sort it out in my head first.

**Comment by iandoug** [comment]
As Arno say, we're getting to the point of diminishing returns. I'm down to the "7 counts" in the unknown words, and the Capitalised ones are starting to look a bit suspicious, so will stop with that process. If we can maybe do another 200 regular words, then I will drop the rest of the unknown (and probably rare) words and make final 'cleaner' corpus.

Metrics:
Found 4,884,656 bad lines out of 26,329,994 lines
Found 1,600,172 distinct bad words and 654,631 distinct good words 
These will improve further.

100 unknown words:

abstenciones	1451
expareja	1434
detenerlo	1420
demostrada	1399
leerse	1395
suene	1378
mediocampo	1373
expandirse	1370
efectuarse	1368
respetada	1362
encaminadas	1361
cuidada	1346
usarla	1342
percatarse	1342
intervenida	1330
lesionada	1321
postularse	1319
deberse	1312
publicarse	1304
denominarse	1294
sustituida	1292
cerrarse	1281
implementadas	1273
repetida	1269
emplearse	1266
evitarse	1254
confundirse	1254
reemplazada	1251
improvisada	1240
tenerse	1238
estudiada	1228
convertirla	1226
ayudarla	1222
pactada	1211
infectada	1209
alzarse	1199
escaparse	1195
interesarse	1194
situaci	1192
quirúrgicamente	1184
romperse	1174
estudiadas	1165
sofisticada	1152
expresidenta	1152
tuiteó	1144
rígida	1142
altísima	1141
dlares	1136
detonaciones	1121
resolverlo	1120
frustraciones	1115
demostrarlo	1112
decirnos	1110
conflictiva	1109
alargada	1108
atribuida	1107
reproducirse	1100
reiterada	1099
hipotética	1098
brindada	1097
reportada	1097
utilizarla	1090
pernoctaciones	1085
ejecutarse	1080
remitida	1079
apreciarse	1079
sumarle	1078
idónea	1075
hospitalizada	1074
contemplada	1071
preguntarnos	1068
trasladarlo	1068
coronarse	1064
provocaciones	1064
íntegra	1060
satisfacciones	1060
pagarse	1058
entrenarse	1055
fructífera	1054
endémica	1049
deteriorada	1046
desarrollándose	1045
divulgada	1042
implementada	1034
llevarle	1034
comprarlo	1022
inundadas	1016
celebrarlo	1014
situacin	1012
depositada	1010
indiscriminada	1007
llevarnos	1006
destacarse	1004
enviarle	1002
permitirle	996
deliberaciones	994
goleadora	991
basarse	989
connotaciones	989
gustaria	987

Thanks, Ian

**Reply by Lobo-Feroz** [reply]
Hi there,

As the last batch, the majority of these words are valid, though a little uncommon, with a few typos in between:

situaci --> situación
dlares --> dólares 
situacin --> situación
gustaria --> gustaría

The rest are OK.

**Reply by iandoug** [reply]
Thanks, have updated and rerunning.

**Comment by binarybottle** [comment]
@iandoug & @Lobo-Feroz -- thank you so much for your persistence and diligence!

**Comment by iandoug** [comment]
Last batch, thanks.
Current metrics:
Found 4775454 bad lines out of 26329994 lines
Found 1599906 distinct bad words and 654893 distinct good words 

I looked at the Gutenburg code, it pulls all books .. for English, I discarded any that had non-US-ANSI chars. That won't work here. Instead I got a book list which shows language, will parse that then use wget/curl to pull the Spanish ones.

interactúan	980
explicarse	976
escuchada	975
desplegadas	973
incrementarse	973
dejándolo	972
consensuada	970
traumática	970
travs	967
discutida	966
oacute	966
centroizquierda	966
escuchadas	965
captada	953
disfrutarlo	946
interpretarse	946
aproximarse	944
facilitada	938
encaminada	931
reñida	926
presentarlo	926
matarla	923
traducirse	919
recibirlo	916
pensarse	916
compartirlo	915
arrebatarle	915
aportada	912
sostenerse	908
parecerse	905
incluirse	901
informarle	897
probarlo	894
conservada	894
asentarse	891
enviarlo	890
absorbida	889
importarle	883
engañosa	880
descalificaciones	876
dividirse	875
pagarlo	875
ltimos	872
salvarle	865
ébola	856
escucharse	855
acompañarlo	855
comandada	854
embarcarse	852
exponencialmente	849
elevarse	848
completarse	840
ininterrumpidamente	839
precampaña	838
minoritaria	836
proclamarse	836
consagrarse	834
buscarse	834
hallarse	834
anotarse	833
solucionarlo	832
concurrida	830
dedicándose	830
vertidas	828
ltimo	825
ansiada	825
adquirirse	825
percibida	824
considerarla	823
centésimas	820
adjudicarse	819
agitada	818
colarse	818
desplegada	817
extraída	816
viralizó	814
progenitora	809
estrellarse	806
realizarle	806
apoyarlo	804
extendiéndose	804
frontalmente	804
cubrebocas	802
lograrse	795
transportada	793
visitarse	791
evaluadas	791
evaluada	790
inflacionaria	790
salvarlo	786
estacionada	785
obtenerse	784
emprendidas	782
descargarse	776
cruzarse	775
estrenarse	774
modificarse	774
apreciaciones	772
protegerlos	771
acomodarse	765

**Reply by Lobo-Feroz** [reply]
Same same. Mostly uncommon words, some typos:

travs --> través
ltimos --> últimos
ltimo --> último

This one is special:

**oacute** 

oacute is not spanish, in fact I think this is html markup to display an **ó**, the complete form being  `&oacute;` but since you've stripped the symbols you are left with oacute.

maybe in the rest of the words you have more like this: aacute eacute iacute uacute ntilde and others

here's a list: https://www.ou.edu/research/electron/internet/special.shtml

**Reply by iandoug** [reply]
Thanks, you are correct about it being HTML. 
I'll first nuke the bad words and then see what sort of mess remains.

**Reply by iandoug** [reply]
Had some other stuff I needed to sort out over our long weekend, hope to carry on with this today .. one more work work thing to sort out first.

**Comment by binarybottle** [comment]
@iandoug -- After years of work, I have created a crowdsource-data-driven, multi-objective approach to optimizing keyboard layouts that I am much happier with. Rather than design based on my own intuitions, I collected data from hundreds of typists about what key-pairs they prefer to type, and empirically derived ergonomics objectives and corresponding key-pair scoring criteria. I am preparing to submit a paper about this work for publication by Halloween (personal deadline) and include a few metrics for evaluation that I developed, but I would like to include some independent online keyboard layout scorer results. In addition to KLAnext, the only online layout scorers I could find are Colemak effort (https://colemakmods.github.io/mod-dh/compare.html), Cyanophage (https://cyanophage.github.io/), Oxey (https://oxey.dev/playground/index.html), and keysolve-web (https://galileoblues.github.io/keysolve-web/). What do you think of these? What do you recommend?

**Reply by binarybottle** [reply]
@iandoug -- I ran across this stats table compiled from Cyanophage. Unless you have a better suggestion, I'll include these data for comparison:
https://getreuer.info/posts/keyboards/alt-layouts/stats.html

**Reply by iandoug** [reply]
Hi Arno
Troubling question. :-). I've been meaning to write a post/article addressing these issues, it will cover items touched on below.

I have major issues with cyanophage, oxey, keysolve, and the bot they use on the Discord server.

Firstly, the physical layouts for the first two are not ANSI/ISO.

The way Enter, Shift, and even Space are handled, if at all by the analyzer, hugely affects various metrics. You can't just ignore them or make assumptions.

Many of the keys are missing. The digits 0, 1, 9 are more common than j, q, z, but are not included.

Then we have WHAT is measured.

They have "skipgrams". I would argue this depends on the typist.

Then there is the focus on "rolls", with the idea that "in-rolls good, out-rolls bad". On the one hand, this ignores the way language works ... eg THougHT ... siNG, iGNore ... one way may be more common, but you will get rolls both ways. Pianists show that there is no speed difference between in or out, so what is the point of the metric?

There is also "redirects" or "pinballs" ... again, from my point of view, this is a made-up metric that does not affect anything. Music relies on redirects to create "hooks", and musicians do not complain about them. It's a non-issue.

Then we have the corpora ... AFAIK they use corpora from one of the typing sites, the Discord bot can use Shai's corpus IIRC, and some others. They work by analysing bigrams or trigrams, but then you need to look at how good the corpus is that they analysed. Did they include all chars on the keyboard? Do they include Caps separately to lower (and thus shift)? Do they include Enter and Space? If not, then the stats are meaningless.

There are other analysers, including other spins on KLA. You can find links here under Layout analysis: https://www.keyboard-design.com/tools.html

See also the [Keyboard Layouts (3rd edition)](https://docs.google.com/document/d/1W0jhfqJI2ueJ2FNseR4YAFpNfsUM-_FlREHbpNGmC2o/edit?tab=t.dg66iecr8la0#heading=h.c9bgu99m6gk6) but with the caveat that it is based on the analyzers above and thus I have major issues with it where the metrics are involved. Same with Pascal's list.

Dario's analyser (link on my Tools page) is good, but you need to decide on which options you like.

Okay, enough for now. May edit later.

**Reply by binarybottle** [reply]
@iandoug -- Thank you for your thoughtful response! 

I hear your concerns. Recognizing that the world of typing metrics is a complex one, I was hoping for a comparison focused on simple, interpretable measures comparing lowercase letters that is independent of capitalization, punctuation, spaces, and minimally affected by ANSI/ISO differences. Given these constrains, which do you recommend? 

My study includes a statistical analysis to determine the degree to which reports of typing preference support a few of the assumptions above. 

I eagerly await any article you end up writing detailing your concerns!

**Reply by binarybottle** [reply]
I should add that I need a defensible argument for why I chose the metrics/analyzers that I did in the publication. So any reasoned selection you can think of would be greatly appreciated! If KLAnext/Dario/others, please make recommendations for appropriate options.

**Comment by binarybottle** [comment]
@iandoug and @Lobo-Feroz -- As mentioned in my comment above, I have developed a new method for generating optimal keyboard layouts in different languages. English and Spanish are almost ready. I wanted to ask you both: do you have any concerns or recommendations based on the work you did to generate the revised Leipzig Spanish corpus or based on the first version of Engram-es?
